tools/revng-convert-ft/Main.cpp

```diff
@@ -0,0 +1,161 @@
+/// \file Main.cpp
+/// \brief This tool is an interface on top of `revng/StaticAnalysis/ABI` used
+/// for easy function type conversion.
+///
+/// It works on `revng::model`. All the valid and convertible functions as
+/// well as all the types they depend on from the input model are added to
+/// the output model
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <fstream>
+#include <streambuf>
+#include <string>
+
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/Signals.h"
+
+#include "revng/Model/Processing.h"
+
+namespace Options {
+
+using namespace llvm::cl;
+
+OptionCategory Category("FT-Converter Options");
+
+enum ConvertionType { Raw, CABI };
+
+auto OpV = values(clEnumValN(Raw,
+                             "raw",
+                             "Convert all the CABI functions to "
+                             "`RawFunctionType` representation"),
+                  clEnumValN(CABI,
+                             "cabi",
+                             "Convert all the Raw functions to "
+                             "`CABIFunctionType` representation"));
+using CT = ConvertionType;
+constexpr const char *OpDesc = "the convertion direction";
+opt<CT> Operation("target", desc(OpDesc), OpV, init(Raw), cat(Category));
+
+using namespace model::ABI;
+using ABI = model::ABI::Values;
+constexpr const char *TrgDesc = "The ABI of the input/output binaries.";
+#define clABIDescription(name) clEnumVal(name, model::ABI::getDescription(name))
+auto TaV = values(clABIDescription(SystemV_x86_64),
+                  clABIDescription(SystemV_x86),
+                  clABIDescription(SystemV_x86_regparm_3),
+                  clABIDescription(SystemV_x86_regparm_2),
+                  clABIDescription(SystemV_x86_regparm_1),
+                  clABIDescription(Microsoft_x86_64),
+                  clABIDescription(Microsoft_x86_64_vectorcall),
+                  clABIDescription(Microsoft_x86_64_clrcall),
+                  clABIDescription(Microsoft_x86_cdecl),
+                  clABIDescription(Microsoft_x86_stdcall),
+                  clABIDescription(Microsoft_x86_thiscall),
+                  clABIDescription(Microsoft_x86_fastcall),
+                  clABIDescription(Microsoft_x86_clrcall),
+                  clABIDescription(Microsoft_x86_vectorcall),
+                  clABIDescription(AArch64),
+                  clABIDescription(ARM),
+                  clABIDescription(SystemV_MIPS_o32),
+                  clABIDescription(SystemZ_s390x));
+opt<ABI> TargetABI("abi", Required, desc(TrgDesc), TaV, cat(Category));
```

Comment:
I'm thinking again about this and you can avoid writing all of them by hand either at compile-time using `make_indexed_sequence` and friends or at run-time as follows:

```c++
template<typename T>
class DynamicOptionList : public opt<T> {
public:
  template<typename... Types>
  DynamicOptionList(Types... Args) : opt<T>(Args...) {
    for (0 .. model::ABI::Count) {
      this->getParser().addLiteralOption(...);
    }
  }
};
DynamicOptionList TargetABI("abi", Required, desc(TrgDesc), cat(Category));
```

Given that we foresee users introducing new ABIs, I wouldn't bother doing the compile-time version.

Reply:
Wait, this is actually a good point - because we cannot possible let user introduce anything into this data-structure! This code is run when global variables are initialized, so we need to technically have the list of all the ABIs at this point and I don't see a way to achieve that without some heavy trickery.

Reply:
I did go with `make_indexed_sequence` approach but won't close this just yet - let's discuss it a bit more.

Reply:
I think this might force us to go with something like `llvm::cl::opt<std::string>` for the ABI name and then parse it manually. Help messages wouldn't be as nice, but I guess it can't be helped.

For now `make_indexed_sequence` should suffice, do you agree?